File: c/quicksort.c
#include <stdio.h>
#include <stdlib.h>

void swap(int* arr, int i, int j) {
  int tmp = arr[i]; arr[i] = arr[j]; arr[j] = tmp;
}

void quicksort(int* arr, int st, int end) {
  if(st == end) return;

  int i, sep = st;
  for(i = st + 1; i < end; i++) {
    if(arr[i] < arr[st]) swap(arr, ++sep, i);
  }

  swap(arr, st, sep);
  quicksort(arr, st, sep);
  quicksort(arr, sep + 1, end);
}

int main() {
  int n; scanf("%d", &n);
  int* arr = malloc(n * sizeof(int));

  int i;
  for(i = 0; i < n; i++)
    scanf("%d", &arr[i]);

  quicksort(arr, 0, n);

  for(i = 0; i < n; i++)
    printf(i == 0 ? "%d" : " %d", arr[i]);
  printf("\n");

  free(arr);
  return 0;
}

File: go/quicksort.go
package main

import "fmt"

func quicksort(arr []int) {
	if len(arr) > 1 {
		sep := 0
		for i := 1; i < len(arr); i++ {
			if arr[i] < arr[0] {
				sep++
				arr[i], arr[sep] = arr[sep], arr[i]
			}
		}

		arr[0], arr[sep] = arr[sep], arr[0]
		quicksort(arr[:sep])
		quicksort(arr[sep+1:])
	}
}

func main() {
	var n int
	fmt.Scanf("%d", &n)

	var arr = make([]int, n)
	for i := 0; i < n; i++ {
		fmt.Scanf("%d", &arr[i])
	}

	quicksort(arr)

	for i := 0; i < n; i++ {
		if i > 0 {
			fmt.Printf(" ")
		}
		fmt.Printf("%d", arr[i])
	}
	fmt.Println()
}
